source/fab/builder.py

```diff
@@ -74,12 +78,48 @@ def __init__(self, config: Config):
             (config.workspace / BUILD_OUTPUT).mkdir()
 
     def run(self):
+        self.init_logging()
+        self.init_metrics()
+
+        artefacts = dict()
+        try:
+            with TimerLogger(f'running {self.config.label} build steps') as steps_timer:
+                for step in self.config.steps:
+                    with TimerLogger(step.name) as step_timer:
+                        step.run(artefacts=artefacts, config=self.config)
+                    send_metric('steps', step.name, step_timer.taken)
+        except Exception as err:
+            raise Exception(f'\n\nError running build steps:\n{err}')
+        finally:
+            self.finalise_metrics(steps_timer.taken)
+
+    def init_logging(self):
         logger.info(f"{datetime.now()}")
-        logger.info(f"use_multiprocessing = {self.config.use_multiprocessing}")
         if self.config.use_multiprocessing:
-            logger.info(f"n_procs = {self.config.n_procs}")
+            logger.info(f'machine cores: {multiprocessing.cpu_count()}')
+            logger.info(f'available cores: {len(os.sched_getaffinity(0))}')
+            logger.info(f'using n_procs = {self.config.n_procs}')
+        logger.info(f"workspace is {self.config.workspace}")
```

Comment:
It would be well worth taking a few moments to nail down how you want to use the different logging levels in this project. We failed to do that for Gung Ho and now everything is either "info" or an error.

How much of this output do you want to see in a normal run and how much should be for debug purposes only?

Reply:
I think that work was done by the wider community before us, and we can stand on their shoulders. The language gives us _debug,_ _info,_ _warning_ and _error_, and I think they're self-explanatory. So it's a debug message if you only want to see it when something's going wrong. It's an info message if you want to see light-weight, general progress through a run. A warning if you want a quieter run.

In our current mode of operation we like clean, concise, info-level logging so we can keep an eye on a run and look back when something's not right. We regularly tweak the logging to show only what's interesting to us. Some of these messages may move to debug in to future, and we might revert the logging level to warning but I guess it's too early to speculate on that.

Reply:
I think you'd be amazed at how many different interpretations of the concept "info" there are but if you are happy with it as is I don't have a problem. As you say, it can all be reorganised later.